README.md
# inkscape-silhouette

[![Run Python tests](https://github.com/fablabnbg/inkscape-silhouette/actions/workflows/run-tests.yml/badge.svg)](https://github.com/fablabnbg/inkscape-silhouette/actions/workflows/run-tests.yml)

An extension to drive a Silhoutte Cameo and similar plotter devices from within inkscape.
100% pure python, ontop of the libusb backend.

Here is the wiki with photos and a video: https://github.com/fablabnbg/inkscape-silhouette/wiki

## Supported Devices

This extension should work with the following devices:

* Silhouette Portrait
* Silhouette Portrait 2 (working confirmed)
* Silhouette Portrait 3
* Silhouette Cameo
* Silhouette Cameo 2
* Silhouette Cameo 3
* Silhouette Cameo 4
* Silhouette Cameo 4 Pro
* Silhouette Curio (partial success confirmed in #36)
* Craft Robo CC200-20
* Craft Robo CC300-20
* Silhouette SD 1
* Silhouette SD 2

## Installation

### Ubuntu 20.10 or newer (features Inkscape 1.0+ natively)

WARNING: SNAP packages may cause issues.

Install Inkscape
* `sudo apt install inkscape`

Install requirements
* `sudo apt install python3-usb`

Install inkscape-silhouette
* https://github.com/fablabnbg/inkscape-silhouette/releases
  Scroll down to Downloads and click on the *.deb file.

### Ubuntu 18.04 or 20.04

WARNING: SNAP packages may cause issues.

Add Inkscape 1.0+ repo
* `sudo add-apt-repository ppa:inkscape.dev/stable`
* `sudo apt-get update`

Install Inkscape
* `sudo apt install inkscape`

Install requirements
* `sudo apt install python3-usb`

Install inkscape-silhouette
* https://github.com/fablabnbg/inkscape-silhouette/releases
  Scroll down to Downloads and click on the *.deb file.

### Other Debian based Linux

* Download https://github.com/fablabnbg/inkscape-silhouette/archive/main.zip
* Unzip the archive into a directory (which will be called inkscape-silhouette-main by default)
* In a terminal, change into that directory
* Execute `make install-local` to install just in your user account, or (if you have permissions) `sudo make install`
to install for all users

* `sudo apt-get install python3-usb` if you have permissions, otherwise `python3 -m pip install usb`
* restart inkscape, check that you see new menu entries "Extensions -> Export -> Send to Silhouette"
and " ... -> Silhouette Multi Action".


### openSUSE

* Same as Debian-based, except install the usb package with `sudo zypper in python-usb`

### Arch Linux

```shell
sudo pacman -S inkscape python-lxml python-pyusb
git clone https://github.com/fablabnbg/inkscape-silhouette.git
cd inkscape-silhouette
```

and then either `make install-local` to install just for your user account, or `sudo make install`

### Mac OS X

* Install prerequisites:
  * install homebrew http://brew.sh/
  * `brew install libusb`
  * `brew install python3`
* Install the extension:
  * `./install_osx.py`
  * Add brew python for user extensions. For details on selecting a specific interpreter version see [Inkscape Wiki - Extension Interpreters](https://wiki.inkscape.org/wiki/index.php/Extension_Interpreters#Selecting_a_specific_interpreter_version_.28via_preferences_file.29):
    * `python-interpreter="/usr/local/bin/python3"` on X86 platform
    * `python-interpreter="/opt/homebrew/bin/python3"` on ARM platform (Apple Silicon)

### FreeBSD

Note the recipe here specifies `py38-libusb1`. In case this is out of date, you need to choose
the usb package appropriate to the version of python that runs by default as `python3`.

```
sudo pkg install inkscape py38-libusb1
cd /tmp
wget -c "https://github.com/fablabnbg/inkscape-silhouette/archive/main.zip"
unzip main.zip
cd inkscape-silhouette-main
sudo make install   # OR: make install-local  # latter installs only for this user
```

### Windows

#### Driver 

These steps must be done with Silhouette device plugged in to USB port.

* Download newest Zadig from http://zadig.akeo.ie/
* Go to menu options `List all devices`
* Look for USB Printing Support in the dropdown list
* Ensure USB ID is: `0B4D` (Graftek America)
* Select driver `libusb-win32 (v1.2.6.0)` which will install a `libusb0`-Port for Windows
* Click replace driver

To later undo:

* Run Zadig again
* Go to menu options `List all devices`
* Look for USB Printing Support in the dropdown list
* Ensure USB ID is: `0B4D` (Graftek America)
* Select driver `WinUsb` which will undo the prior change.
* Click replace driver

#### Python adapter

* Inkscape usually comes with a Version of Python; ensure that feature under `Program Files/Python` is ticked upon installation or change/add features accordingly
* Install pip (a package manager for python):
  * Download `get-pip.py` from https://bootstrap.pypa.io/get-pip.py and copy to the `bin` directory, e.g. `C:\Program Files\Inkscape\bin`
  * Open command line and navigate to the same directory, then enter `.\python.exe '.\get-pip.py'`
* Install pyusb:
  * Still in command line enter `.\python.exe -m pip install pyusb`

#### Silhouette inkscape extension itself

* Download https://github.com/fablabnbg/inkscape-silhouette/archive/main.zip
* Open the downloaded file and select the following five items: `silhouette`, `sendto_silhouette.inx`, `sendto_silhouette.py`, `silhouette_multi.inx`, `silhouette_multi.py`
* Extract them to your `share\inkscape\extensions` directory, e.g. `C:\Program Files\Inkscape\share\inkscape\extensions`
* Restart inkscape

## Usage

1. Open your document with inkscape.
2. Ensure the unit of document width and height is mm or inch, but not px. (File - Document settings - Page - Custom - Unit mm) Otherwise you may observe differences in dimensions at inkscape 0.91/0.92, because default dpi has changed from 90 to 96.
3. Convert all objects and texts to paths (Path - Convert object to path)
4. Select the parts you want to plot.
5. Open the extension. If you want to use the same cut settings for all of the paths in your file, use "Extensions -> Export -> Send to Silhouette." If you want use different cut settings based on the colors of different items in your file, use "Extensions -> Export -> Silhouette Multi Action."
6. In the case of Multi Action, there is a first screen that is primarily for debugging. Typically you can just leave all of the boxes on this unchecked and click "Apply."
7. Set your desired plot parameters. There are numerous aspects you can control with the dialog, here are just the core highlights:
  * **X-Offset, Y-Offset**  An additional offset of your drawing from the top left corner. Default is 0/0
  * **Tool Cut/Pen**        Cut mode drews small circles for orientation of the blade, Pen mode draws exactly as given.
  * **Media**               Select a predfined media or set to custom settings.
  * **Speed**               Custom speed of the movements
  * **Pressure**            Custom Pressure on the blade. One unit is said to be 7g force.

  In Multi Action, you can select the color you want settings to apply to and then set all the same parameters, but with potentially different settings for each color. You can also change the order in which the colors are cut, and uncheck the box in the "Perform Action?" column to ignore a color altogether.
8. To start the cut, in "Send to Silhouette, click the "Apply" button; in "Silhouette Multi" click the "Execute" button.

## Templates
* Templates showing the cutting mat on a background layer can be found in `examples/mat_templates`
* Copy those files into the `templates` subdirectory below inkscapes configuration directory
* To identify the correct path open inkscape's preferences and selecting `System`. There you find the path as `User templates`
* Those templates can then be selected within the dialog available through `File` &rarr; `New from Template...`
* Once you have created a new document from those templates you can import other `*.svg`-files and place the contained objects for cutting

## Troubleshooting

```python
>>> import usb.core
>>> usb.core.find()
<usb.core.Device object at 0xb720fb8c>
>>>
```

If this reports `no usb.core.Device` to you, please help troubleshoot.

```python
python
>>> import usb.core
>>> usb.version_info[0]
```

This fails on win32/64 with 'module has no attribute 'version info' which then causes Graphtec.py to error even though usb.core is installed.

## Using of registration marks

The plotter will search the registration marks at the given positions.
If it founds the marks, they will serve as accurate reference and define the origin.
Therefore it is necessary to set the correct offset values of the mark.
As a result the cut will go precisely along the graphics.


To plot with registration marks do the following:

1. Open the document which fit to your setup, e.g. `examples/registration-marks-cameo-silhouette-a4-maxi.svg` for Silhouette Cameo using A4 paper format.
2. Insert your cutting paths and graphics on the apropriate layers.
3. Printout the whole document including registration marks. You probably want to hide the cutting layer.
4. Select your cutting paths in the document, but exclude regmarks and graphics.
5. On the **Regmarks** tab:
  * Check **Document has registration marks**
  * Check **Search for registration marks**
6. Set all following parameters according to the registration file used:
  * **X mark distance** (e.g. *190*)
  * **Y mark distance** (e.g. *277*)
  * **Position of regmark from document left** (e.g. *10*)
  * **Position of regmark from document top** (e.g. *10*)
7. Start cut.

On some devices have an offset between the search optics and the cutting knife.
For enhanced precision, you may have to set an offset on **X-Offset** and/or **Y-Offset** on the **Silhouette** tab to compensate.

## Features

* Path sorting for monotonic cut. We limit backwards movement to only a few
  millimeters, and make the knive pull only towards sharp corners
  so that most designs can be done without a cutting mat!
* Coordinate system conforms to inkscape SVG.
* Exact Margins. Can start at (0,0).
* Pen mode used to avoid the precut movement of the knive.
  Those movements are visible a) at the left hand side, when
  starting, b) at each sharp turn.
* Bounding Box. Can optionally plot (or calculate only)
  the bounding box instead of plotting all strokes.
  This can be used (with low pressure=1 or removed knive) to just
  check, where the plot would be.
* The standalone script `arrow_test.py` can be used to test drive
  the `SilhoutteCameo` class.
* Robust communication with the device. Small writes and timeouts are
  handled gracefully. Timeouts will occur, when we travel far with low speed.
* Multipass: Can repeat each stroke multiple times to enhance plot or
  cut quality. This can also be used to attempt a cut without cutting mat, by
  applying very little pressure.
* reverse toggle options, to cut the opposite direction. This might also be
  helpful with mat-free cutting via multipass.
* honors hidden layers.

## Misfeatures of InkCut that we do not 'feature'

* object transforms are missing most of the time.
* Stars, polygons, and boxes are plotted not closed, the final stroke
  is missing. (Must be me, no?)
* always plots all layers, even if hidden.

## TODO

* Implement the triangle in a square test cut.

* test MatFree cutting strategy with the WC-Wunderbach-Wimpern font, which is especially
  well suited as a test-case.

* improve MatFree cutting by finding a better scan sort algorithm.
  Wide shadow casting towards negative y?

* Implement paper-zip as a seperate inkscape extension.

## References

* http://wiki.inkscape.org/wiki/index.php/Extensions
* http://wiki.inkscape.org/wiki/index.php/INX_Parameters
* http://wiki.inkscape.org/wiki/index.php/ExtensionsSystem


HISTORY.txt
# code snippets visited to learn the extension 'effect' interface:
# - http://sourceforge.net/projects/inkcut/
# - http://code.google.com/p/inkscape2tikz/
# - http://wiki.inkscape.org/wiki/index.php/PythonEffectTutorial
# - http://github.com/jnweiger/inkscape-gears-dev
# - http://code.google.com/p/eggbotcode/
# - http://www.bobcookdev.com/inkscape/better_dxf_output.zip
#
# Porting to OSX
# - https://github.com/pmonta/gerber2graphtec/blob/master/file2graphtec
# - https://github.com/pmonta/gerber2graphtec/blob/master/README
#
# 2013-05-09 jw, V0.1 -- initial draught
# 2013-05-10 jw, V0.2 -- can plot simple cases without transforms.
# 2013-05-11 jw, V0.3 -- still using inkcut/plot.py -- fixed write(),
#                        improved logging, flipped y-axis.
# 2013-05-12 jw, v0.4 -- No unintended multipass when nothing is selected.
#                        Explicit multipass option added.
#                        Emplying recursivelyTraverseSvg() from eggbotcode
#                        TODO: coordinate system of page is not exact.
# 2013-05-13 jw, v0.5 -- transporting docWidth/docHeight to dev.plot()
# 2013-05-15 jw, v0.6 -- Replaced recursivelyTraverseSvg() and friends with the
#                        versions from eggbot.py, those from eggbot_hatch.py
#                        would only do closed paths. Makes sense for them, but
#                        not for us.
#                        Added no_device=True debugging aid to SilhouetteCameo()
# 2013-05-17 jw, v0.7 -- Honor layer visibility by checking style="display:none"
#                        penUP()/penDown() bugfix to avoid false connections between objects.
#                        Added option reversetoggle, to cut the opposite direction.
# 2013-05-19 jw, v0.8 -- Split GUI into two pages. Added dummy and mat-free checkboxes.
#                        misc/corner_detect.py done, can now load a dump saved by dummy.
#                        Udev rules and script added, so that we get a nice notify
#                        guiding users towards inkscape, when connecting a device.
# 2013-05-25 jw, v0.9 -- mat_free option added. The slicing and sharp corner strategy
#                        appears useful.
# 2013-05-26 jw, v1.0 -- Some tuning done. fixed preset scaling, improved path recombination.
# 2013-05-26 jw, v1.1 -- Strategy.MatFree.path_overshoot() added. With 0.2mm overshoot
#                        the paper now comes apart almost by itself. great.
#                        Buffer percent estimation added. We now have an estimate
#                        how long the buffered data will need.
# 2013-05-30 jw, v1.2 -- Option autocrop added. Speed improvement: only parse visible layers.
# 2013-05-31 jw, v1.3 -- sharp_turn() now takes self.sharp_turn_fwd_ratio parameter.
#                        test_drive.py now draws arrows. All [0], [1] converted to new .x, .y syntax.
#                        Split Geometry.py from Strategy.py; class Barrier implemented.
# 2013-10-24 jw, v1.4 -- Fixed an abort in Strategy. when pt has no seg.
# 2013-11-02 jw, v1.5 -- Added protability code. This might eventually lead to
#                        working code on windows and macosx too. Still linux only.
# 2013-11-08 jw, v1.6 -- supporting mm in getLength().
# 2013-12-16 jw, v1.7 -- https://github.com/jnweiger/inkscape-silhouette/issues/1
#                        fixed. Silly copy/paste bug. Looks like I miss a testsuite.
# 2014-01-23 jw, v1.8 -- improving portability by using os.devnull, os.path.join, tempfile.
#                        Partial fixes for https://github.com/jnweiger/inkscape-silhouette/issues/2
#                        Enumerating devices if none are found.
# 2014-01-28 jw, v1.9 -- We cannot expect posix semantics from windows.
#                        Experimental retry added when write returns 0.
#                        issues/2#issuecomment-33526659
# 2014-02-04 jw, v1.9a -- new default: matfree false, about page added.
# 2014-03-29 jw, v1.9b -- added own dir to sys.path for issue#3.
# 2014-04-06 jw, v1.9c -- attempted workaround for issue#4
# 2014-07-18 jw, v1.9d -- better diagnostics. hints *and* (further down) a stack backtrace.
# 2014-09-18 jw, v1.10 -- more diagnostics, fixed trim margins aka autocrop to still honor hardware margins.
# 2014-10-11 jw, v1.11 -- no more complaints about empty <text/> elements. Ignoring <flowRoot>
# 2014-10-25 jw, v1.12 -- better error messages.
# 2014-10-31 jw, v1.13 -- fixed usb.core.write() without interface parameter. Set Graphtec.py/need_interface if needed.
# 2015-06-06 jw, v1.14 -- fixed timout errors, refactored much code.
#                         Added misc/silhouette_move.py misc/silhouette_cut.py, misc/endless_clock.py
# 2016-01-15 jw, v1.15 -- ubuntu loads the wrong usb library.
# 2016-05-15 jw, v1.16 -- merged regmarks code from https://github.com/fablabnbg/inkscape-silhouette/pull/23
# 2016-05-17 jw, v1.17 -- fix avoid dev.reset in Graphtec.py, fix helps with
#                         https://github.com/fablabnbg/inkscape-silhouette/issues/10
# 2016-05-21 jw, v1.18 -- warn about python-usb < 1.0 and give instructions.
#                         Limit pressure to 18. 19 or 20 make the machine
#                         scroll forward backward for several minutes.
#                         Support document unit inches. https://github.com/fablabnbg/inkscape-silhouette/issues/19
# 2016-12-18 jw, v1.19 -- support for dashed lines added. Thanks to mehtank
#                         https://github.com/fablabnbg/inkscape-silhouette/pull/33
#                         Added new cutting strategy "Minimized Traveling"
#                         Added parameter for blade diameter
# 2018-06-01 jw, v1.20 -- Make it compile again. Hmm.
# 2019-07-25 jw, v1.21 -- merge from github.com/olegdeezus/inkscape-silhouette
#                         merge from fablabnbg
# 2019-08-03 jw, v1.22 -- added a copy of pyusb-1.0.2 as a fallback on any platform.

Graphtec.py
# (c) 2013 jw@suse.de
# driver for a Graphtec Silhouette Cameo plotter.
# modelled after https://github.com/nosliwneb/robocut.git 
#
import sys, time, usb.core
    
# taken from 
#  robocut/CutDialog.ui
#  robocut/CutDialog.cpp

MEDIA = [
# media, pressure, speed, cap-color, name
  ( 100,   27,     10,  "yellow", "Card without Craft Paper Backing"),
  ( 101,   27,     10,  "yellow", "Card with Craft Paper Backing"),
  ( 102,   10,     10,  "blue",   "Vinyl Sticker"),
  ( 106,   14,     10,  "blue",   "Film Labels"),
  ( 111,   27,     10,  "yellow", "Thick Media"),
  ( 112,    2,     10,  "blue",   "Thin Media"),
  ( 113,   10,     10,  "pen",    "Pen"),
  ( 120,   30,     10,  "blue",   "Bond Paper 13-28 lbs (105g)"),
  ( 121,   30,     10,  "yellow", "Bristol Paper 57-67 lbs (145g)"),
  ( 122,   30,     10,  "yellow", "Cardstock 40-60 lbs (90g)"),
  ( 123,   30,     10,  "yellow", "Cover 40-60 lbs (170g)"),
  ( 124,    1,     10,  "blue",   "Film, Double Matte Translucent"),
  ( 125,    1,     10,  "blue",   "Film, Vinyl With Adhesive Back"),
  ( 126,    1,     10,  "blue",   "Film, Window With Kling Adhesive"),
  ( 127,   30,     10,  "red",    "Index 90 lbs (165g)"),
  ( 128,   20,     10,  "yellow", "Inkjet Photo Paper 28-44 lbs (70g)"),
  ( 129,   27,     10,  "red",    "Inkjet Photo Paper 45-75 lbs (110g)"),
  ( 130,   30,      3,  "red",    "Magnetic Sheet"),
  ( 131,   30,     10,  "blue",   "Offset 24-60 lbs (90g)"),
  ( 132,    5,     10,  "blue",   "Print Paper Light Weight"),
  ( 133,   25,     10,  "yellow", "Print Paper Medium Weight"),
  ( 134,   20,     10,  "blue",   "Sticker Sheet"),
  ( 135,   20,     10,  "red",    "Tag 100 lbs (275g)"),
  ( 136,   30,     10,  "blue",   "Text Paper 24-70 lbs (105g)"),
  ( 137,   30,     10,  "yellow", "Vellum Bristol 57-67 lbs (145g)"),
  ( 138,   30,     10,  "blue",   "Writing Paper 24-70 lbs (105g)"),
  ( 300, None,   None,  "custom", "Custom"),
]

#  robocut/Plotter.h:53 ff
VENDOR_ID_GRAPHTEC = 0x0b4d 
PRODUCT_ID_CC200_20 = 0x110a
PRODUCT_ID_CC300_20 = 0x111a
PRODUCT_ID_SILHOUETTE_SD_1 = 0x111c
PRODUCT_ID_SILHOUETTE_SD_2 = 0x111d
PRODUCT_ID_SILHOUETTE_CAMEO =  0x1121
PRODUCT_ID_SILHOUETTE_PORTRAIT = 0x1123

DEVICE = [
 { 'vendor_id': 0x0b4d, 'product_id': 0x1123, 'name': 'Silhouette Portrait', 
   'width_mm':  203, 'length_mm': 3000, 'regmark': True },
 { 'vendor_id': 0x0b4d, 'product_id': 0x1121, 'name': 'Silhouette Cameo',    
   # margin_top_mm is just for safety when moving backwards with thin media
   # margin_left_mm is a physical limit, but is relative to width_mm!
   'width_mm':  304, 'length_mm': 3000, 'margin_left_mm':9.0, 'margin_top_mm':5.0, 'regmark': True },
 { 'vendor_id': 0x0b4d, 'product_id': 0x110a, 'name': 'Craft Robo CC200-20', 
   'width_mm':  200, 'length_mm': 1000, 'regmark': True },
 { 'vendor_id': 0x0b4d, 'product_id': 0x111a, 'name': 'Craft Robo CC300-20' },
 { 'vendor_id': 0x0b4d, 'product_id': 0x111c, 'name': 'Silhouette SD 1' },
 { 'vendor_id': 0x0b4d, 'product_id': 0x111d, 'name': 'Silhouette SD 2' },
]

def _bbox_extend(bb, x, y):
    # The coordinate system origin is in the top lefthand corner.
    # Downwards and rightwards we count positive. Just like SVG or HPGL.
    # Thus lly is a higher number than ury
    if not 'llx' in bb or x < bb['llx']: bb['llx'] = x
    if not 'urx' in bb or x > bb['urx']: bb['urx'] = x
    if not 'lly' in bb or y > bb['lly']: bb['lly'] = y
    if not 'ury' in bb or y < bb['ury']: bb['ury'] = y

class SilhouetteCameo:
  def __init__(self, log=sys.stderr):
    """ This initializer simply finds the first known device.
        The default paper alignment is left hand side for devices with known width 
        (currently Cameo and Portrait). Otherwise it is right hand side. 
        Use setup() to specify your needs.
    """
    self.leftaligned = False            # True: only works for DEVICE with known hardware.width_mm
    self.log = log
    dev = None
    for hardware in DEVICE:
      dev = usb.core.find(idVendor=hardware['vendor_id'], idProduct=hardware['product_id'])
      if dev:
        dev.hardware = hardware
        break

    if dev is None:
      dev = usb.core.find(idVendor=VENDOR_ID_GRAPHTEC)
      dev.hardware = { 'name': 'Unknown Graphtec device' }

    if dev is None:
      raise ValueError('No Graphtec Silhouette devices found. Check USB and Power')
    print >>self.log, "%s found on usb bus=%d addr=%d" % (dev.hardware['name'], dev.bus, dev.address)

    if dev.is_kernel_driver_active(0):
      print >>self.log, "is_kernel_driver_active(0) returned nonzero"
      if dev.detach_kernel_driver(0):
        print >>self.log, "detach_kernel_driver(0) returned nonzero"
    dev.reset();

    dev.set_configuration()
    try:
      dev.set_interface_altsetting()      # Probably not really necessary.
    except usb.core.USBError:
      pass
    self.dev = dev
    self.regmark = False                # not yet implemented. See robocut/Plotter.cpp:446
    if 'width_mm' in self.dev.hardware: self.leftaligned = True 

  def write(s, string, timeout=1000):
    """Send a command to the device. Long commands are sent in chunks of 4096 bytes.
       A nonblocking read() is attempted before write(), to find spurious diagnostics."""

    # robocut/Plotter.cpp:73 says: Send in 4096 byte chunks. Not sure where I got this from, I'm not sure it is actually necessary.
    try:
      resp = s.read(timeout=10) # poll the inbound buffer
      if resp:
        print >>s.log, "response before write('%s'): '%s'" % (string, resp)
    except:
      pass
    endpoint = 0x01
    chunksz = 4096
    r = 0
    o = 0
    msg=''
    while o < len(string):
      if o and s.log:
        s.log.write(" %d%% %s\r" % (100.*o/len(string),msg))
        s.log.flush()
      chunk = string[o:o+chunksz]
      try:
        r = s.dev.write(endpoint, string[o:o+chunksz], interface=0, timeout=timeout) 
      except Exception as e:
        # raise USBError(_str_error[ret], ret, _libusb_errno[ret])
        # usb.core.USBError: [Errno 110] Operation timed 
        # print >>s.log, "Write Exception: %s, %s errno=%s" % (type(e), e, e.errno)
        import errno
        if e.errno == errno.ETIMEDOUT:
          time.sleep(1)
          msg += 't'
          continue
      else:
        if len(msg):
          msg = ''
          s.log.write("\n")

      # print >>s.log, "write([%d:%d], len=%d) = %d" % (o,o+chunksz, len(chunk), r)
      if r <= 0:
        raise ValueError('write %d bytes failed: r=%d' % (len(chunk), r))
      o += r
    if o != len(string):
      raise ValueError('write all %d bytes failed: o=%d' % (len(string), o))
      
  def read(s, size=64, timeout=5000):
    """Low level read method"""
    endpoint = 0x82
    data = s.dev.read(endpoint, size, timeout=timeout, interface=0) 
    if data is None:
      raise ValueError('read failed: none')
    return data.tostring()

  def status(s):
    """Query the device status. This can return one of the three strings
       'ready', 'moving', 'unloaded' or a raw (unknown) byte sequence."""
    # Status request.
    s.write("\x1b\x05")
    resp = s.read(timeout=5000)
    if resp[-1] != '\x03': raise ValueError('status response not terminated with 0x03: %s' % (resp[-1]))
    if resp[:-1] == '0': return "ready"
    if resp[:-1] == '1': return "moving"
    if resp[:-1] == '2': return "unloaded"
    return resp[:-1]
  
  def initialize(s):
    """Send the init command. Called by setup()."""
    # taken from robocut/Plotter.cpp:331 ff
    # Initialise plotter.
    s.write("\x1b\x04")

  def home(s):
    """Send the home command. Untested. Called by setup()."""
    s.write("TT\x03")

  def get_version(s):
    """Retrieve the firmware version string from the device."""
    s.write("FG\x03")
    resp = s.read(timeout=10000) # Large timeout because the plotter moves.
    return resp[0:-2]   # chop of 0x03

  def setup(s, media=132, speed=None, pressure=None, pen=None, trackenhancing=False, landscape=False, leftaligned=None):
    """media range is [100..300], default 132, "Print Paper Light Weight"
       speed range is [1..10], default None, from paper (132 -> 10)
       pressure range is [1..33], default None, from paper (132 -> 5)
          Notice: Cameo runs trackenhancing if you select a pressure of 19 or more.
       pen: True or False, default None (media dependant)
       trackenhancing: True or False, default False (setting ignored??)
       landscape: True or False, default False
       leftaligned: Loaded media is aligned left(=True) or right(=False), default: device dependant
    """

    s.initialize()
    s.home()

    if leftaligned is not None:
      s.leftaligned = leftaligned
    if media is not None: 
      if media < 100 or media > 300: media = 300
      s.write("FW%d\x03" % media);
      if pen is None: 
        if media == 113:
          pen = True
        else:
          pen = False
      for i in MEDIA:
        if i[0] == media: 
          print >>s.log, "Media=%d, cap='%s', name='%s'" % (media, i[3], i[4])
          if pressure is None: pressure = i[1]
          if    speed is None:    speed = i[2]
    if speed is not None: 
      if speed < 1: speed = 1
      if speed > 10: speed = 10
      s.write("!%d\x03" % speed);
    if pressure is not None: 
      s.write("FX%d\x03" % pressure);
      # s.write("FX%d,0\x03" % pressure);       # oops, graphtecprint does it like this
    if s.leftaligned:
      print >>s.log, "Loaded media is expected left-aligned."
    else:
      print >>s.log, "Loaded media is expected right-aligned."

    # robocut/Plotter.cpp:393 says:
    # // I think this sets the distance from the position of the plotter
    # // head to the actual cutting point, maybe in 0.1 mms (todo: Measure blade).
    # // It is 0 for the pen, 18 for cutting.
    # // C possible stands for curvature. Not that any of the other letters make sense...
    cutter = 18
    if pen: cutter = 0
    s.write("FC%d\x03" % cutter)

    if trackenhancing is not None: 
      if trackenhancing: 
        s.write("FY1\x03")
      else:
        s.write("FY0\x03")

    if landscape is not None: 
      if landscape: 
        s.write("FN1\x03")
      else:
        s.write("FN0\x03")

    # // No idea what this does.
    s.write("FE0\x03")

    # // Again, no idea. Maybe something to do with registration marks?
    s.write("TB71\x03")
    resp = s.read(timeout=10000)        # // Allow 10s. Seems reasonable.
    if resp != "    0,    0\x03":
      raise ValueError("setup: Invalid response from plotter.")

  def cut_bbox(s, cut):
    """Find the bouding box of the cut, returns (xmin,ymin,xmax,ymax)"""
    bb = {}
    for path in cut:
      for pt in path:
        _bbox_extend(bb,pt[0],pt[1])
    return bb

  def flip_cut(s, cut):
    """this returns a flipped copy of the cut about the x-axis, 
       keeping min and max values as they are."""
    bb = s.cut_bbox(cut)
    new_cut = []
    for path in cut:
      new_path = []
      for pt in path:
        new_path.append((pt[0], bb['lly']+bb['ury']-pt[1]))
      new_cut.append(new_path)
    return new_cut

  def mirror_cut(s, cut):
    """this returns a mirrored copy of the cut about the x-axis, 
       keeping min and max values as they are."""
    bb = s.cut_bbox(cut)
    new_cut = []
    for path in cut:
      new_path = []
      for pt in path:
        new_path.append((bb['llx']+bb['urx']-pt[0], pt[1]))
      new_cut.append(new_path)
    return new_cut

  def page(s, mediawidth=210.0, mediaheight=297.0, margintop=None, marginleft=None, cut=None, offset=None, bboxonly=None):
    """cut is a list of paths. A path is a sequence of 2-tupel, all measured in mm.
           The tool is lowered at the beginning and raised at the end of each path.
       offset = (X_MM, Y_MM) can be specified, to easily move the design to the 
           desired position.  The top and left media margin is always added to the 
           origin. Default: margin only.
       bboxonly:  True for drawing the bounding instead of the actual cut design; 
                  False for not moving at all (just return the bounding box). 
                  Default: None for normal cutting.
       Example: The letter Y can be represented with 
                cut=[[(0,0),(4.5,10),(4.5,20)],[(9,0),(4.5,10)]]
    """
    bbox = { }
    clipcount = 0
    total = 0
    if margintop  is None and 'margin_top_mm'  in s.dev.hardware: margintop  = s.dev.hardware['margin_top_mm']
    if marginleft is None and 'margin_left_mm' in s.dev.hardware: marginleft = s.dev.hardware['margin_left_mm']
    if margintop  is None: margintop = 0
    if marginleft is None: marginleft = 0

    if s.leftaligned and 'width_mm' in s.dev.hardware:
      # marginleft += s.dev.hardware['width_mm'] - mediawidth  ## FIXME: does not work.
      mediawidth =   s.dev.hardware['width_mm']

    print >>s.log, "mediabox: (%g,%g)-(%g,%g)" % (mediawidth,mediaheight,marginleft,margintop)

    # // Begin page definition.
    s.write("FA\x03")   # query someting?
    try:
      resp = s.read(timeout=100)
      if len(resp) > 1:
        print >>s.log, "FA: '%s'" % (resp[:-1])
    except:
      pass

    width  = int(0.5+20.*mediawidth)
    height = int(0.5+20.*mediaheight)
    top    = int(0.5+20.*margintop)
    left  = int(0.5+20.*marginleft)
    if width < left: width=left
    if height < top:  height=top

    x_off = left
    y_off = top
    if offset is not None:
      if type(offset) != type([]) and type(offset) != type(()):
        offset = (offset, 0)
      x_off += int(.5+offset[0]*20.)
      y_off += int(.5+offset[1]*20.)
    # print >>s.log, "x_off=%d, y_off=%d" % (x_off,y_off)

    s.write("FU%d,%d\x03" % (height-top, width-left))
    s.write("FM1\x03")          # // ?
    if s.regmark:
      raise ValueError("regmark code not impl. see robocut/Plotter.cpp:446")
    else:
      s.write("TB50,1\x03")     #; // ???

    # // I think this is the feed command. Sometimes it is 5588 - maybe a maximum?
    s.write("FO%d\x03" % (height-top))

    p = "&100,100,100,\\0,0,Z%d,%d,L0" % (width,height)

    for path in cut:
      if len(path) < 2: continue
      # x = path[len(path)-1][0]*20. + x_off
      # y = path[len(path)-1][1]*20. + y_off
      x = path[0][0]*20. + x_off
      y = path[0][1]*20. + y_off
      _bbox_extend(bbox, x,y)
      total += 1
    
      last_inside = True
      if x < left:  
        x = left
        last_inside = False
      if x > width: 
        x = width
        last_inside = False
      if y < top:    
        y = top
        last_inside = False
      if y > height: 
        y = height
        last_inside = False
      if not last_inside: clipcount += 1

      if bboxonly is None:
        p += ",M%d,%d" % (int(0.5+width-x), int(0.5+y))
      # for j in reversed(range(0,len(path)-1)):
      for j in range(1,len(path)):
        x = path[j][0]*20. + x_off
        y = path[j][1]*20. + y_off
        _bbox_extend(bbox, x,y)
        total += 1

        inside = True
        if x < left:
          x = left
          inside = False
        if x > width:
          x = width
          inside = False
        if y < top:
          y = top
          inside = False
        if y > height:
          y = height
          inside = False
        if not inside: clipcount += 1

        if bboxonly is None:
          if inside and last_inside:
            p += ",D%d,%d" % (int(0.5+width-x), int(0.5+y))
          else:
            # // if outside the range just move
            p += ",M%d,%d" % (int(0.5+width-x), int(0.5+y))
        last_inside = inside

    if bboxonly == True:
      # move the bouding box
      p += ",M%d,%d" % (int(0.5+width-bbox['llx']), int(0.5+bbox['ury']))
      p += ",D%d,%d" % (int(0.5+width-bbox['urx']), int(0.5+bbox['ury']))
      p += ",D%d,%d" % (int(0.5+width-bbox['urx']), int(0.5+bbox['lly']))
      p += ",D%d,%d" % (int(0.5+width-bbox['llx']), int(0.5+bbox['lly']))
      p += ",D%d,%d" % (int(0.5+width-bbox['llx']), int(0.5+bbox['ury']))

    p += "&1,1,1,TB50,0\x03"   #; // TB maybe .. ah I dunno. Need to experiment. No idea what &1,1,1 does either.
    s.write(p)
    s.write("FO0\x03")          # // Feed the page out.
    s.write("H,")               # // Halt?

    return { 
        'bbox': bbox, 
        'clipcount': clipcount,
        'total': total,
        'unit' : 1/20.,
        'mbox': { 'urx':width, 'ury':top, 'llx':left, 'lly':height }
      }

